tests/test_requirements.py

```diff
@@ -0,0 +1,343 @@
+
+    req_info = next(req_iter)
+    assert isinstance(req_info, ReqInfo)
+    assert "GoodRequirement" == req_info.project_name
+
+    req_info = next(req_iter)
+    assert isinstance(req_info, ReqInfo)
+    assert "AnotherRequirement" == req_info.project_name
+
+    with pytest.raises(ParseError) as exc_info:
+        next(req_iter)
+
+    assert exc_info.value.logical_line == LogicalLine(
+        raw_text=(
+            "BadRequirement\\\n"
+            "[extra, another]; \\\n"
+            'bad_marker == "2.7" \\\n'
+            "    --global-option=foo # End of line comment.\n"
+        ),
+        processed_text='BadRequirement[extra, another]; bad_marker == "2.7" --global-option=foo',
+        source=other_requirement_file,
+        start_line=7,
+        end_line=10,
+    )
+
+
+class MarkerWithEq(Marker):
+    @classmethod
+    def wrap(cls, marker):
+        # type: (Optional[Marker]) -> Optional[MarkerWithEq]
+        return None if marker is None else MarkerWithEq(str(marker))
+
+    def __eq__(self, other):
+        return type(other) == MarkerWithEq and str(self) == str(other)
+
+
+def req(
+    project_name=None,  # type: Optional[str]
+    url=None,  # type: Optional[str]
+    marker=None,  # type: Optional[str]
+    editable=False,  # type: bool
+    is_local_project=False,  # type: bool
+):
+    # type: (...) -> ReqInfo
+    return ReqInfo(
+        line=None,
+        project_name=project_name,
+        url=url,
+        marker=MarkerWithEq.wrap(marker),
+        editable=editable,
+        is_local_project=is_local_project,
+    )
+
+
+def normalize_results(req_infos):
+    # type: (Iterable[Union[Constraint, ReqInfo]]) -> List[Union[Constraint, ReqInfo]]
+    def normalize_req_info(req_info):
+        return req_info._replace(line=None)._replace(marker=MarkerWithEq.wrap(req_info.marker))
+
+    return [
+        normalize_req_info(req_info)
+        if isinstance(req_info, ReqInfo)
+        else Constraint(normalize_req_info(req_info.req_info))
+        for req_info in req_infos
+    ]
+
+
+def test_parse_requirements_stress(chroot):
+    # type: (str) -> None
+    with safe_open(os.path.join(chroot, "other-requirements.txt"), "w") as fp:
+        fp.write(
+            # This includes both example snippets taken directly from
+            # https://pip.pypa.io/en/stable/reference/pip_install/#requirements-file-format
+            # not already covered by
+            # https://pip.pypa.io/en/stable/reference/pip_install/#example-requirements-file.
+            dedent(
+                """\
+                SomeProject
+                SomeProject == 1.3
+                SomeProject >=1.2,<2.0
+                SomeProject[foo, bar]
+                SomeProject~=1.4.2
+                
+                SomeProject ==5.4 ; python_version < '2.7'
+                SomeProject; sys_platform == 'win32'
+                
+                SomeProject @ file:///somewhere/over/here
```

Comment:
While you're at it, consider testing PEP 440 URLs, in addition to files. We use these in our Pants docs:

```
Django@ git+https://github.com/django/django.git
Django@ git+https://github.com/django/django.git@stable/2.1.x
Django@ git+https://github.com/django/django.git@fd209f62f1d83233cc634443cfac5ee4328d98b8
```

Reply:
OK, well the existing test _does_ have a direct URL reference and one URL is as good as another here ... but added anyhow.